src/Deinterleave.cpp

```diff
@@ -198,202 +384,345 @@ class Deinterleaver : public IRGraphMutator {
+        std::vector<int> input_vector_lane_counts;
+        input_vector_lane_counts.reserve(op->lanes);
+        std::fill_n(std::back_inserter(input_vector_lane_counts),
+                    /*num=*/op->lanes, /*value=*/op->value.type().lanes());
+
+        // Shuffle indices are specified as-if the src vector
+        // is self-concatenated `op->lanes` times.
+        // Decompose that into repetition index and lane within src vector.
+        std::vector<DecomposedIndice> decomposed_input_lanes =
+            decompose_shuffle_indices(input_vector_lane_counts, lanes.indices);
+
+        // Which repetitions of the source vector are actually demanded?
+        IndiceXFormMapping repetition_xform = IndiceXFormMapping::get(
+            decomposed_input_lanes, op->lanes,
+            [](const DecomposedIndice &i) { return i.outer; });
+
+        // Which lanes of the source vector are demanded *at all*?
+        IndiceXFormMapping input_lane_xform = IndiceXFormMapping::get(
+            decomposed_input_lanes, op->value.type().lanes(),
+            [](const DecomposedIndice &i) { return i.inner; });
+
+        Expr in;
+        // Now that we know which lanes of the src vector are demanded, recurse.
+        {
+            ScopedValue<DemandedLanes> old_lanes(lanes,
+                                                 input_lane_xform.indices);
+            in = mutate(op->value);
+        }
+
+        // Refresh vector lane count knowledge after mutating source vector.
+        input_vector_lane_counts.clear();
+        std::fill_n(std::back_inserter(input_vector_lane_counts),
+                    /*num=*/repetition_xform.indices.size(),
+                    /*value=*/input_lane_xform.indices.size());
+
+        // Now that we (might have) dropped some of the repetitions
+        // of the source vector, and maybe mutated said source vector,
+        // update the decomposed shuffle mask.
+        for (DecomposedIndice &i : decomposed_input_lanes) {
+            i.outer = *repetition_xform.translation[i.outer];
+            i.inner = *input_lane_xform.translation[i.inner];
+        }
+
+        // Now that we are done with mutation, let's reconstruct the indices.
+        std::vector<int> new_lanes = recompose_shuffle_indices(
+            input_vector_lane_counts, decomposed_input_lanes);
+
+        // Now, we can recreate the hopefully-smaller Broadcast.
+        if (int lanes = repetition_xform.indices.size(); lanes != 1) {
+            in = Broadcast::make(in, lanes);
         }
-        if (op->value.type().lanes() > 1) {
-            // There is probably a more efficient way to this.
-            return mutate(flatten_nested_ramps(op));
+
+        // But, it is possible that we *still* don't demand all of the produced
+        // lanes, in which case we need a Shuffle after all.
+        if (in.type().lanes() != new_lanes.size()) {
+            in = Shuffle::make({in}, new_lanes);
```

Comment:
(We only get here if the source had more than one lane)
Conceptual question: do we really want to keep the `broadcast`
even if we will then immediately drop some of the produced lanes?

Reply:
I think the complexity here is due to deinterleave running before flattening nested vectors. So the question is if you ask for the even lanes of:

```
ramp(ramp(0, 1, 10), broadcast(100), 10)
```

Should you get:

```
ramp(ramp(0, 1, 10), broadcast(200), 5)
```

Or:

```
ramp(ramp(0, 2, 5), broadcast(100), 10)
```

And I guess the correct answer is the latter, which is why things are complicated.

So in the case you're talking about, I guess it's a shuffle of a broadcast of vectors, e.g. taking the even lanes of:

```
broadcast(ramp(0, 1, 3), 3)
```

i.e. [0, 1, 2, 0, 1, 2, 0, 1, 2], which is I guess [0, 2, 1, 0, 2]

yuck.

Maybe we can just do the correct thing here, even if it produces ugly IR, and rely on the simplifier to clean up any cases that can be cleaned up